includes/class-wcsatt-product.php

```diff
@@ -0,0 +1,856 @@
+			self::set_product_property( $product, 'subscription_period_interval', null );
+			self::set_product_property( $product, 'subscription_length', null );
+
+			$scheme_set = true;
+		}
+
+		do_action( 'wcsatt_set_product_subscription_scheme', $key, $active_scheme_key, $product );
+
+		return $scheme_set;
+	}
+
+	/**
+	 * Set the product as purchasable on a recurring basis only.
+	 *
+	 * @param  WC_Product  $product                 Product object to set.
+	 * @param  boolean     $is_forced_subscription  Value.
+	 */
+	public static function set_forced_subscription( $product, $is_forced_subscription ) {
+		self::set_product_property( $product, 'has_forced_subscription', $is_forced_subscription ? 'yes' : 'no' );
+	}
+
+	/*
+	|--------------------------------------------------------------------------
+	| Helpers
+	|--------------------------------------------------------------------------
+	*/
+
+	/**
+	 * Property getter (compatibility wrapper).
+	 *
+	 * @param  WC_Product  $product   Product object.
+	 * @param  string      $property  Property name.
+	 * @return mixed
+	 */
+	private static function get_product_property( $product, $property ) {
+
+		if ( WCS_ATT_Core_Compatibility::is_wc_version_gte_2_7() ) {
+			$value = $product->get_meta( '_' . $property, true );
+		} else {
+			$value = isset( $product->$property ) ? $product->$property : '';
+		}
+
+		return $value;
+	}
+
+	/**
+	 * Property setter (compatibility wrapper).
+	 *
+	 * @param  WC_Product  $product  Product object.
+	 * @param  string      $name     Property name.
+	 * @param  string      $value    Property value.
+	 * @return mixed
+	 */
+	private static function set_product_property( $product, $name, $value ) {
+
+		if ( WCS_ATT_Core_Compatibility::is_wc_version_gte_2_7() ) {
+			$product->add_meta_data( '_' . $name, $value, true );
+		} else {
+			$product->$name = $value;
+		}
+	}
+
+	/*
+	|--------------------------------------------------------------------------
+	| Filters
+	|--------------------------------------------------------------------------
+	*/
+
+	/**
+	 * Hooks onto 'woocommerce_is_subscription' to trick Subs into thinking it is dealing with a subscription.
+	 *
+	 * @param  boolean     $is
+	 * @param  int         $product_id
+	 * @param  WC_Product  $product
+	 * @return boolean
+	 */
+	public static function filter_is_subscription( $is, $product_id, $product ) {
+
+		if ( ! $product ) {
+			return $is;
+		}
+
+		if ( self::is_subscription( $product ) ) {
+			$is = true;
+		}
+
+		return $is;
+	}
+
+	/**
+	 * Delete object meta in use by the application layer.
+	 *
+	 * @param  WC_Product  $product
+	 */
+	public static function delete_reserved_meta( $product ) {
+		$reserved_meta_keys = array( 'has_forced_subscription', 'subscription_schemes', 'active_subscription_scheme_key', 'default_subscription_scheme_key' );
+		foreach ( $reserved_meta_keys as $reserved_meta_key ) {
+			$product->delete_meta_data( '_' . $reserved_meta_key );
+		}
```

Comment:
> This would definitely work. We could map data to objects using object instance IDs. It still feels more natural to me that object-specific data should be stored on the object.

Yeah, I can understand that.

One advantage (or disadvantage depending on usage) of storing it is `WCS_ATT_Product` is that the same data will be available for other instances of the product. For example, if some other code only has the product ID, and instantiates a new instance of that product, should it also get the same `has_forced_subscription`, `subscription_schemes`, `active_subscription_scheme_key` and `default_subscription_scheme_key` values as the other instance or instances of the product? 

Answering that will answer where the data belongs IMO.